README.md
# Welcome to ballgownR's home!

## What is this for?

* This repository is the home for the ballgownR package as well as tests related to it's development.
* Once the package has matured, it will move to a repo completely dedicated to sharing it. Meaning, without any of initial tests and explorations.

## What is ballgown?

To fill in later

## Installation instructions:

```S
install.packages("devtools")

library(devtools)
install_github("clickme", "nachocab")
install_github("ballgownR-devel", "lcolladotor", subdir="ballgownR")
```


## Visualization example:

Check /test-viewGene for example code. Here are some examples:

* [toy-test](http://htmlpreview.github.io/?https://github.com/lcolladotor/ballgownR-devel/blob/master/test-viewGene/toy-test.html)
* [toy-test-2](http://htmlpreview.github.io/?https://github.com/lcolladotor/ballgownR-devel/blob/master/test-viewGene/toy-test-2.html)
* [toy-test-3](http://htmlpreview.github.io/?https://github.com/lcolladotor/ballgownR-devel/blob/master/test-viewGene/toy-test-3.html)
* [toy-test-4](http://htmlpreview.github.io/?https://github.com/lcolladotor/ballgownR-devel/blob/master/test-viewGene/toy-test-4.html)

Powered by [clickme](https://github.com/nachocab/clickme)


ballgownR/NEWS
CHANGES IN VERSION 1.0.2
------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o viewGene() is now split into two functions: infoGene() and the plotting function viewGene().
    o Examples (along with example data) are now available for all functions.
    
NEW FEATURES

    o viewGene() will now show a title and color code.
    o infoGene() now allows the user to specify which count information to show and whether to use the intron data or not. 

BUG FIXES

    o No changes classified as 'bug fixes'
    
CHANGES IN VERSION 1.0.1
------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o readGown() now assign full sample names instead of 1 to n

NEW FEATURES

    o viewGene() allows you to visualize the results from readGown() using an interactive html

BUG FIXES

    o No changes classified as 'bug fixes'

CHANGES IN VERSION 1.0.0
------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o readGown() is now available
	
NEW FEATURES

    o readGown() allows you to read in ballgown's output

BUG FIXES

    o No changes classified as 'bug fixes'


clickme-test/test_line2.R
library(clickme)
library(colorspace)

### Custom line

## set path
set_root_path("/Users/lcollado/enigma2/ballgownR-devel/clickme-test/")



mammals_path <- system.file(file.path("examples", "one_zoom", "data", "mammals.tree"), package="clickme")
cities <- c("Boston", "NYC", "Philadelphia")
n <- 30
df2 <- data.frame(line = rep(cities, each = n), x = rep(1:n, length(cities)), y = c(sort(rnorm(n)), -sort(rnorm(n)),sort(rnorm(n))))

## Initial manual color test
col <- '["#393b79","#5254a3","#6b6ecf"]'
clickme(df2, "line_with_focus2", params=list(color=col), html="test.html")

## Test with groups
test <- data.frame(line=paste0("veryLongText", rep(letters[1:9], each=30)), x=rep(1:30, 9), y=rnorm(30*9))
col2 <- paste0('["', paste0(rep(rainbow_hcl(3, start = 30, end = 300), each=3), collapse='","'), '"]')
clickme(test, "line_with_focus2", params=list(color=col2), html="test2.html")

## Testing whether you can show an "exon"
col <- '["#393b79","#5254a3","#6b6ecf","#6b6ecf"]'
df3 <- rbind(df2, data.frame(line=rep("exon", 5), x=11:15, y=-5))
clickme(df3, "line_with_focus2", params=list(color=col), html="test3.html")


## without specifying color param
clickme(df3, "line_with_focus2", html="test4.html")

ballgownR/R/infoGene.R
#' Data extraction of the ballgown results to visualize with \code{viewGene}
#'
#' \code{infoGene} extracts the information according to a geneID. This can be the rcounts or mrcounts per exon (and introns optionall). Alternatively, it can extract the raw coverage.
#'
#' @param geneID specifies the geneID to look at
#' @param gown specifies the output from \code{readGown}
#' @param group specifies the sample groups and will be used to determine the colors.
#' @param countIntron specifies whether to use the intron data or not. If set to TRUE, \code{whichCount} has to be mrcount, rcount or ucount.
#' @param whichCount has to be mrcount, rcount or ucount if \code{countIntron=TRUE}. Otherwise it can also be cov and mcov.
#' @return A list with the information needed to make the plot with \code{viewGene}
#' @export
#' @author Leonardo Collado-Torres \email{lcollado@@jhsph.edu}
#' @examples
#' dataDir <- system.file("extdata", "ballgownData", package="ballgownR")
#' samplePattern <- "sample"
#' gown <- readGown(dataDir=dataDir, samplePattern=samplePattern)
#' info <- system.file("extdata", "sample_info.txt", package="ballgownR")
#' info <- read.table(info, header=TRUE)
#' match <- sapply(names(gown$dirs), function(x) { which(info$dir == x)})
#' geneInfo <- infoGene(geneID="gene_1", gown=gown, group=info$outcome[match])
#' @seealso viewGene

infoGene <- function(geneID, gown, group, countIntron=TRUE, whichCount="mrcount") {
	## Find region of interest
	idx.t <- which(gown$trans$gene_id == geneID)
	
	## Find start and end
	start <- min(gown$trans$start[idx.t])
	end <- max(gown$trans$end[idx.t])	
	nBases <- end - start + 1
	
	## Subset trans, exon and intron
	t.id <- gown$trans$t_id[idx.t]
	names(t.id) <- gown$trans$t_name[idx.t]
	e.id <- lapply(t.id, function(x) { gown$e2t$e_id[ gown$e2t$t_id == x ] })
	
	## Get sample names and the exon/intron rcount columns
	count.e <- which(gsub("\\..*", "", names(gown$exon[e.id[[1]][1], ])) == whichCount)
	samples <- gsub(paste0(whichCount, "\\."), "", names(gown$exon[e.id[[1]][1], ]))[count.e]
	
	## Get exon information by transcript
	exons.df <- lapply(1:length(e.id), function(e) {	
		## Get transcript name
		line <- names(e.id)[[e]]		
		res <- lapply(e.id[[e]], function(exon) {	
			x <- c(gown$exon$start[exon]-1, gown$exon$start[exon], gown$exon$end[exon], gown$exon$end[exon]+1)			
			y <- c(0, e, e, 0)
			data.frame(line=rep(line, length(x)), x=x, y=y)				 
		})
		do.call(rbind, res)
	})
	exons.df <- do.call(rbind, exons.df)
	
	## Build data: either summarized data 
				
	## Initialize rcount data
	count.df <- data.frame(line=rep(samples, each=nBases), x=start:end, y=rep(NA, each=nBases))
	
	## Count exon information
	count.df <- .countInfo(ids=e.id, df=gown$exon, whichCols=count.e, result=count.df)
	
	## Count intron information
	if(countIntron & whichCount %in% c("ucount", "rcount", "mrcount")) {
		i.id <- lapply(t.id, function(x) { gown$i2t$i_id[ gown$i2t$t_id == x ] })
		count.i <- which(gsub("\\..*", "", names(gown$intron[i.id[[1]][1], ])) == whichCount)
		count.df <- .countInfo(ids=i.id, df=gown$intron, whichCols=count.i, result=count.df)
	}
	
	## Finish
	toAdd <- count.df

	## Purge useless info
	toAdd.filt <- .filterInfoGene(toAdd)
	
	## Merge data
	data <- list(exons=exons.df, transInfo=toAdd.filt, start=start, end=end, group=group, geneID=geneID, ylab=whichCount)
	
	## Done
	return(data)
	
}

## This function counts the exon/intron information
.countInfo <- function(ids, df, whichCols, result) {
	for(id in 1:length(ids)) {
		for(xon in ids[[id]]) {
			x <- seq(df$start[xon], df$end[xon])			
			vals <- as.vector(as.matrix(df[xon, whichCols]))
			y <- rep(vals, each=length(x))
			idx <- result$x %in% x
			result$y[idx] <- pmin(result$y[idx],  y, na.rm=TRUE)
		}
	}
	return(result)
}

## This function removes uninformative points to avoid overloading the html
.filterInfoGene <- function(data) {
	## The filtering is done by unique elements of data$line (for example, by sample)
	result <- lapply(unique(data$line), function(x) {
		sub <- data[ data$line == x, ]
		which.diff <- which(diff(sub$y) != 0)
		if(length(which.diff) == 0) {
			res <- rbind(sub[1, ], sub[nrow(sub), ])
		} else {
			res <- rbind(sub[1, ], sub[as.vector(sapply(which.diff, function(z) { c(z, z+1)})), ])
			res <- rbind(res, sub[nrow(sub), ])
		}
		return(res)
	}) 
	result <- do.call(rbind, result)
	rownames(result) <- 1:nrow(result)
	return(result)
}

ballgownR/R/readGown.R
#' Read ballgown output *.ctab files from a data directory with multiple samples
#'
#' \code{readGown} reads the *.ctab files that ballgown creates. It then uses plyr to merge the information from multiple samples.
#'
#' @param dataDir must be a character that specifies the data directory path. Inside of it there must be a directory for each sample.
#' @param samplePattern must be a character that specifies the pattern to use that identifies the sample directories
#' @param verbose is a logical. If TRUE updates will be printed via message.
#' @return a list with e2t: mappting between exons and transcripts, i2t: mapping between introns and transcripts, intron: i_data.ctab merged in wide format, exon: e_data.ctab merged in mide format, trans: t_data.ctab merged in wide format, dirs: sample directories (specifies the order in which they are merged), mergedDate: date this function was run
#' @export
#' @author Leonardo Collado-Torres \email{lcollado@@jhsph.edu}
#' @examples
#' #Read toy data
#' dataDir <- system.file("extdata", "ballgownData", package="ballgownR")
#' samplePattern <- "sample"
#' gown <- readGown(dataDir=dataDir, samplePattern=samplePattern)

#### Main function
readGown <- function(dataDir, samplePattern, verbose=TRUE) {
	if(verbose) message(date())
	
	## Load required pkgs
	suppressMessages(library(plyr))
	
	if(FALSE){
		# for testing
		dataDir <- "/amber2/scratch/lcollado/brain_rna/ballgown"
		samplePattern <- "orb"
	}
	
	## Identify the sample directories
	dirs <- list.files(path=dataDir, pattern=samplePattern, full.names=TRUE)
	names(dirs) <- list.files(path=dataDir, pattern=samplePattern)
	n <- length(dirs)
	
	if(FALSE){
		# for testing
		n <- 3 
	}
	
	## Read linking tables
	if(verbose) message("Reading linking tables")
	e2t <- read.table(list.files(dirs[1], "e2t.ctab", full.names=TRUE), header=TRUE, sep="\t", colClasses=c("integer", "integer"))
	i2t <- read.table(list.files(dirs[1], "i2t.ctab", full.names=TRUE), header=TRUE, sep="\t", colClasses=c("integer", "integer"))
	
	## Order by transcript id
	e2t <- e2t[order(e2t$t_id), ]
	i2t <- i2t[order(i2t$t_id), ]
	rownames(e2t) <- 1:nrow(e2t)
	rownames(i2t) <- 1:nrow(i2t)
	
	## Read counts for all introns in <reference_transcripts>
	if(verbose) message("Reading intron data files")
	intronFiles <- sapply(dirs, list.files, pattern="i_data.ctab", full.names=TRUE)
	intronAll <- lapply(intronFiles, .readIntron)
	
	## Merge the results
	if(verbose) message("Merging intron data")
	intron <- join_all(intronAll, by=c("i_id", "chr", "strand", "start", "end"), type="left")
	colnames(intron)  <- c("i_id", "chr", "strand", "start", "end", paste(c("rcount", "ucount", "mrcount"), rep(names(dirs), each=3), sep="."))
	

	## Read read counts and raw coverage info for all exons in <reference_transcripts>
	if(verbose) message("Reading exon data files")
	exonFiles <- sapply(dirs, list.files, pattern="e_data.ctab", full.names=TRUE)
	exonAll <- lapply(exonFiles, .readExon)
	
	## Read exon data
	if(verbose) message("Merging exon data")
	exon <- join_all(exonAll, by=c("e_id", "chr", "strand", "start", "end"), type="left")
	colnames(exon) <- c("e_id", "chr", "strand", "start", "end", paste(c("rcount", "ucount", "mrcount", "cov", "cov_sd", "mcov", "mcov_sd"), rep(names(dirs), each=7), sep="."))
	
	
	## Read transcript data
	if(verbose) message("Reading transcription data files")
	transFiles <- sapply(dirs, list.files, pattern="t_data.ctab", full.names=TRUE)
	transAll <- lapply(transFiles, .readTrans)
	
	## Merge the results
	if(verbose) message("Merging transcription data")
	trans <- join_all(transAll, by=c("t_id", "chr", "strand", "start", "end", "t_name", "num_exons", "length", "gene_id", "gene_name"), type="left")
	colnames(trans) <- c("t_id", "chr", "strand", "start", "end", "t_name", "num_exons", "length", "gene_id", "gene_name", paste(c("cov", "FPKM"), rep(names(dirs), each=2), sep="."))
	
	if(verbose) message("Wrapping up the results")
	result <- list(e2t=e2t, i2t=i2t, intron=intron, exon=exon, trans=trans, dirs=dirs, mergedDate=date())
	
	if(verbose) message(date())
	# Done!
	return(result)
}

#### Auxiliary functions

## Read intron files
.readIntron <- function(file){
	intron <- read.table(file, header=TRUE, sep="\t", colClasses=c("integer", "character", "factor", "integer", "integer", "integer", "integer", "numeric"))
	intron <- intron[order(intron$i_id), ]
	rownames(intron) <- 1:nrow(intron)
	return(intron)
}

## Read counts and raw coverage
.readExon <- function(file) {
	exon <- read.table(file, header=TRUE, sep="\t", colClasses=c("integer", "character", "factor", "integer", "integer", "integer", "integer", "numeric", "numeric", "numeric", "numeric", "numeric"))
	exon <- exon[order(exon$e_id), ]
	rownames(exon) <- 1:nrow(exon)
	return(exon)
}

## Read transcript data files
.readTrans <- function(file) {
	trans <- read.table(file, header=TRUE, sep="\t", colClasses=c("integer", "character", "factor", "integer", "integer", "character", "integer", "integer", "character", "character", "numeric", "numeric"))
	trans <- trans[order(trans$t_id), ]
	rownames(trans) <- 1:nrow(trans)
	return(trans)
}


generate-test-data/data.R
### Ballgown data

setwd("../ballgownR/inst/extdata")

maindir <- "ballgownData"
sampledirs <- paste0("sample", 1:4)

e2t <- data.frame("e_id"=c(1:3, 1, 3, 1, 2, 2, 3, 4, 5), "t_id"=c(1, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5))
i2t <- data.frame("i_id"=c(1:2, 3, 1, 2, 4), "t_id"=c(1, 1, 2, 3, 4, 5))

edata <- data.frame("e_id"=1:5, "chr"=rep(1, 5), "strand"=rep(c("+", "-"), c(3, 2)), "start"=c(1, 4000, 8000, 3000, 6000), "end"=c(2000, 6000, 12000, 5000, 8000), "rcount"=rep(0, 5), "ucount"=rep(0, 5), "mrcount"=rep(0, 5), "cov"=rep(0, 5), "cov_sd"=rep(0, 5), "mcov"=rep(0, 5), "mcov_sd"=rep(0, 5))

idata <- data.frame("i_id"=1:4, "chr"=rep(1, 4), "strand"=rep(c("+", "-"), c(3, 1)), "start"=c(2001, 6001, 2001, 5001), "end"=c(3999, 7999, 7999, 5999), "rcount"=rep(0, 4), ucount=rep(0, 4), mrcount=rep(0, 4))

tdata <- data.frame("t_id"=1:5, "chr"=rep(1, 5), "strand"=rep(c("+", "-"), c(4, 1)), "start"=rep(c(1, 4000, 3000), c(3, 1, 1)), "end"=rep(c(12000, 8000, 12000, 8000), c(2, 1, 1, 1)), "t_name"=paste0("trans_", 1:5), "num_exons"=rep(c(3, 2), c(1, 4)), "length"=c(8002, 6001, 4001, 6002, 5001), "gene_id"=paste0("gene_", rep(c(1, 2), c(4, 1))), "gene_name"=paste0("name_", rep(c(1, 2), c(4, 1))), "cov"=rep(0, 5), "FPKM"=rep(0, 5))

system(paste("mkdir -p", maindir))

## For now, just easy uniform data
populateData <- function(s) {
	switch(s, 
		"sample1" = 10,
		"sample2" = 20,
		"sample3" = 15,
		"sample4" = 25
	) 
}

for(s in sampledirs) {
	dir <- paste(maindir, s, sep="/")
	system(paste("mkdir -p", dir))
	
	## Write regular files
	write.table(e2t, file=paste(dir, "e2t.ctab", sep="/"), quote=FALSE, sep="\t", row.names=FALSE)
	write.table(i2t, file=paste(dir, "i2t.ctab", sep="/"), quote=FALSE, sep="\t", row.names=FALSE)
		
	## Populate data and write it
	enew <- edata
	inew <- idata
	tnew <- tdata
	
	enew$rcount <- enew$ucount <- enew$mrcount <- rep(populateData(s), nrow(enew))
	set.seed(20130412)
	enew$cov <- enew$mcov <- round(runif(n=nrow(enew), min = populateData(s) - 2, max = populateData(s) + 2))
	inew$rcount <- inew$ucount <- inew$mrcount <- rep(populateData(s), nrow(inew))
	tnew$cov <- tnew$FPKM <- rep(populateData(s), nrow(tnew))
	
	write.table(enew, file=paste(dir, "e_data.ctab", sep="/"), quote=FALSE, sep="\t", row.names=FALSE)
	write.table(inew, file=paste(dir, "i_data.ctab", sep="/"), quote=FALSE, sep="\t", row.names=FALSE)
	write.table(tnew, file=paste(dir, "t_data.ctab", sep="/"), quote=FALSE, sep="\t", row.names=FALSE)
}


### Sample info
info <- data.frame("sample"=paste0("sampleName", 1:4), dir=paste0("sample", 1:4), outcome=rep(c("control", "case"), each=2))
write.table(info, file="sample_info.txt", quote=FALSE, sep="\t", row.names=FALSE)


test-viewGene/test-viewGene-toydata.R
### Check that infoGene() and viewGene() work

suppressMessages(library(ballgownR))
dataDir <- system.file("extdata", "ballgownData", package="ballgownR")
samplePattern <- "sample"
gown <- readGown(dataDir=dataDir, samplePattern=samplePattern)

### Test the visualization
info <- system.file("extdata", "sample_info.txt", package="ballgownR")
info <- read.table(info, header=TRUE)
match <- sapply(names(gown$dirs), function(x) { which(info$dir == x)})
geneInfo <- infoGene(geneID="gene_1", gown=gown, group=info$outcome[match])

## Visualize
viewGene(geneInfo=geneInfo, html="toy-test.html", spacing=0.1, width=1000)

## Coverage
geneInfo2 <- infoGene(geneID="gene_1", gown=gown, group=info$outcome[match], countIntron=FALSE, whichCount="cov")
viewGene(geneInfo=geneInfo2, html="toy-test-2.html", spacing=0.1)

## No intron
geneInfo3 <- infoGene(geneID="gene_1", gown=gown, group=info$outcome[match], countIntron=FALSE)
viewGene(geneInfo=geneInfo3, html="toy-test-3.html", spacing=0.1)

## top
viewGene(geneInfo=geneInfo, html="toy-test-4.html", spacing=0.1, location="top")

###


### Checking infoGene
geneID <- "gene_1"
whichCount <- "mrcount"
countIntron <- TRUE
group <- info$outcome[match]

test-viewGene/genome_info/template/test-translator.R
context("translate genome_info")

test_that("data frames are translated to the format expected by the template", {
    data <- data.frame(line = rep(c("l1", "l2"), each = 2), x = 1:4, y = 11:14)
    expected_data <- list(
                          list(key = "l1",
                               values = list(
                                             list(x = 1, y = 11),
                                             list(x = 2, y = 12)
                                        )
                          ),
                          list(key = "l2",
                               values = list(
                                             list(x = 3, y = 13),
                                             list(x = 4, y = 14)
                                        )
                          )
                     )

	# I'm testing get_data_as_nested_structure instead of get_data_as_json because it's easier to understand expected_data as a list, than as a JSON string.
	nested_data <- get_data_as_nested_structure(data)
	expect_equal(nested_data, expected_data)
})




ballgownR/R/viewGene.R
#' Interactive visualization of the ballgown results
#'
#' \code{viewGene} plots the information extracted with \code{infoGene} in an interactive html
#'
#' @param geneInfo is the output of \code{infoGene}
#' @param geneID specifies the geneID to look at
#' @param exon.color is the color used for the exon lines
#' @param location has to be 'bottom' or 'top' and it controls where the exon tracks are showed
#' @param spacing is the factor by which the exons are separated
#' @param html specifies the html output file
#' @param wdir specifies the directory where the plot will be made. You must have writing permission.
#' @param browse=TRUE if you want to open a browser window to view the graphic
#' @param title specifies the title in the HTML file
#' @param main specifies the title of the plot in the HTML file
#' @param height specifies in pixels the height of the plot.
#' @param width specifies in pixels the width of the plot.
#' @return The html file name of the resulting plot.
#' @export
#' @author Leonardo Collado-Torres \email{lcollado@@jhsph.edu}
#' @examples
#' dataDir <- system.file("extdata", "ballgownData", package="ballgownR")
#' samplePattern <- "sample"
#' gown <- readGown(dataDir=dataDir, samplePattern=samplePattern)
#' info <- system.file("extdata", "sample_info.txt", package="ballgownR")
#' info <- read.table(info, header=TRUE)
#' match <- sapply(names(gown$dirs), function(x) { which(info$dir == x)})
#' geneInfo <- infoGene(geneID="gene_1", gown=gown, group=info$outcome[match])
#' viewGene(geneInfo=geneInfo, html="toy-test.html", spacing=0.1)
#' viewGene(geneInfo=geneInfo, html="toy-test-2.html", spacing=0.1, location="top")
#' @seealso infoGene

viewGene <- function(geneInfo, exon.color="#000000", location="bottom", spacing=0.02, html=NULL, wdir=NULL, browse=TRUE, title=NULL, main=NULL, height=500, width=800) {
	## Load required libraries
	require(clickme)
	require(colorspace)
	
	## for testing
	if(FALSE) {
		exon.color <- "#000000"
		location <- "bottom"
		spacing <- 0.02
		html <- NULL
		wdir <- "/Users/lcollado/enigma2/ballgownR-devel/test-viewGene"
		browse=TRUE
		title <- NULL
		main <- NULL
		height <- 500
		width <- 800
	}
	
	## Assign information form geneInfo
	geneID <- geneInfo$geneID
	group <- geneInfo$group
	exons.df <- geneInfo$exons
	toAdd <- geneInfo$transInfo
	start <- geneInfo$start
	end <- geneInfo$end
	ylab <- geneInfo$ylab
	
	## Set working directory
	if(is.null(wdir)) wdir <- getwd()
	
	## Set default html file name
	if(is.null(html)) html <- paste0(runif(1, max=1000000), "-", geneID, ".html")
		
	## Set default title
	if(is.null(title)) title <- paste("Viewing gene", geneID, "with ballgownR")
	
	## Set default main
	if(is.null(main)) main <- paste("Viewing", ylab, "for gene", geneID, "via <a href='https://github.com/lcolladotor/ballgownR-devel/tree/master/ballgownR'>ballgownR</a>")

	## Copy layout to working directory
	if(!"genome_info" %in% dir()) {
		layout.dir <- system.file(package="ballgownR", "extdata", "genome_info")
		## TODO: make this work on Windows
		system(paste0("cp -r ", layout.dir, " ", wdir, "/"))
	}
	
	## Change clickme's path
	set_root_path(wdir)
	
	## Contruct the colors for the exons
	exons.col <- rep(exon.color, length(unique(exons.df$line)))
	
	## Define sample colors
	group.col <- rainbow_hcl(length(unique(group)), start = 30, end = 300)
	sample.col <- rep(NA, length(group))
	for(j in 1:length(unique(group))){
		k <- unique(group)[j]
		sample.col[group %in% k] <- group.col[k]
	}
	
	## Adjust where to show the exons and it's spacing
	exons.df$y <- exons.df$y * diff(range(toAdd$y)) * spacing
	if(location == "bottom") {
		exons.df$y <- (-1) * exons.df$y * diff(range(toAdd$y)) * spacing
		border.y <- min(c(exons.df$y, toAdd$y)) * 1.1
	} else {
		exons.df$y <- max(toAdd$y) * 1.1 + exons.df$y * diff(range(toAdd$y)) * spacing
		border.y <- max(c(exons.df$y, toAdd$y)) * 1.1
	}
	
	## Set a border	to help visualize the exons
	border <- data.frame(line=c("border", "border"), x=c(start, end), y=c(border.y, border.y))
	
	## Merge data
	data <- rbind(exons.df, toAdd, border)
	
	## Format the colors
	colors <- c(exons.col, sample.col, exons.col[1])
	## Format for js
	colors <- paste0('["', paste0(colors, collapse='","'), '"]')
	
	## Color labels
	group.col.label <- sapply(1:length(unique(group)), function(x) {
		paste0("<font color='", group.col[x], "'>", unique(group)[x], "</font>")
	})
	group.col.label <- paste0("<p>Color code: ", paste(group.col.label, collapse=", "), ".</p>")
	
		
	## Visualize
	result <- clickme(data, "genome_info", params=list(color=colors, title=title, main=main, groupLabel=group.col.label, height=height, width=width), html_file_name=html)
	
	## Done
	return(invisible(result))
	
}